### PR TITLE
Add basic ICMPv6 reply

### PR DESCRIPTION
**Pull request by dlrobertson**
- ICMPv6
  - Handle ICMPv6 echo requests
  - Send ICMPv6 error messages
     - When know listening UDP socket is found in process_udp
     - When the next header is not known
  - Update icmpv6 test to be more useful
- ICMPv4
  - Handle one more case where we are sending too much data

Dependecies: #152

### CODE REVIEW

File: src/iface/ethernet.rs [hunk 1]
```diff
@@ -679,6 +703,36 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
         }
     }
 
+    #[cfg(feature = "proto-ipv6")]
+    fn process_icmpv6<'frame>(&self, _sockets: &mut SocketSet, ip_repr: IpRepr,
+                              ip_payload: &'frame [u8]) -> Result<Packet<'frame>>
+    {
+        let icmp_packet = Icmpv6Packet::new_checked(ip_payload)?;
+        let checksum_caps = self.device_capabilities.checksum.clone();
```
Comment (batonius) [inline src/iface/ethernet.rs]: Unrelated to the PR, but do we really need to clone `self.device_capabilities.checksum` everywhere?

Comment (whitequark) [reply src/iface/ethernet.rs]: Borrow checker and closures.

File: src/iface/ethernet.rs [hunk 2]
```diff
@@ -608,10 +616,26 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
...
+            _ => {
+                // Send back as much of the original payload as we can.
+                //
+                // ICMPv6 Contains no protocol unreachable reason for the
+                // destination unreachable message. Instead a parameter
+                // problem message is returned with the code set to
+                // "unrecognized next header" and the pointer set to the
+                // offset within the invoking packet.
+                let payload_len = cmp::min(ip_payload.len(),
+                                           IPV6_MIN_MTU - ipv6_repr.buffer_len() * 2 - 8);
```
Comment (hjr3) [inline src/iface/ethernet.rs]: This formula is repeated four times (across IPv4 and IPv6). Should it be a function so it is not repeated?

File: src/iface/ethernet.rs [hunk 3]
```diff
@@ -666,9 +690,9 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
                 Ok(Packet::None),
 
             _ => {
-                // Send back as much of the original payload as we can
-                let payload_len = cmp::min(
-                    ip_payload.len(), self.device_capabilities.max_transmission_unit);
+                // Send back as much of the original payload as we can.
+                let payload_len = cmp::min(ip_payload.len(),
+                                           IPV4_MIN_MTU - ipv4_repr.buffer_len() * 2 - 8);
```
Comment (hjr3) [inline src/iface/ethernet.rs]: This looks like a bug fix for the existing implementation. Do we need to update existing tests to ensure this behavior stays correct?

File: src/iface/ethernet.rs [hunk 4]
```diff
@@ -679,6 +703,36 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
...
+        match icmp_repr {
+            // Respond to echo requests.
+            Icmpv6Repr::EchoRequest { ident, seq_no, data } => {
+                let icmp_reply_repr = Icmpv6Repr::EchoReply {
+                    ident:  ident,
+                    seq_no: seq_no,
+                    data:   data
+                };
+                match ip_repr {
+                    IpRepr::Ipv6(ipv6_repr) => Ok(self.icmpv6_reply(ipv6_repr, icmp_reply_repr)),
```
Comment (hjr3) [inline src/iface/ethernet.rs]: Why create `icmp_reply_repr` if it is only used in one of the match arms?

File: src/iface/ethernet.rs [hunk 5]
```diff
@@ -679,6 +703,36 @@ impl<'b, 'c> InterfaceInner<'b, 'c> {
...
+                    _ => Err(Error::Unrecognized),
+                }
+            }
+
+            // Ignore any echo replies.
+            Icmpv6Repr::EchoReply { .. } => Ok(Packet::None),
+
+            // FIXME: do something correct here?
+            _ => Err(Error::Unrecognized),
```
Comment (hjr3) [inline src/iface/ethernet.rs]: According to [IPv6 Essentials](https://www.safaribooksonline.com/library/view/ipv6-essentials/0596001258/):

> If a node receives an ICMPv6 informational message of unknown type, it must be silently discarded.

File: src/iface/ethernet.rs (hunk 2 above)
Comment (dlrobertson) [reply src/iface/ethernet.rs]: probably

File: src/iface/ethernet.rs (hunk 3 above)
Comment (dlrobertson) [reply src/iface/ethernet.rs]: I already have one for UDP... Not sure if one for this case is necessary

File: src/iface/ethernet.rs (hunk 4 above)
Comment (dlrobertson) [reply src/iface/ethernet.rs]: Yeah, probably not necessary. I just copied ICMPv4.

File: src/iface/ethernet.rs (hunk 5 above)
Comment (dlrobertson) [reply src/iface/ethernet.rs]: Good to know. There are some known ICMPv6 error messages that could hit this. I think this FIXME is largely to pass ICMPv6 error messages to higher layers. (An issue should probably be created for this)

**Comment by dlrobertson** [comment]
Updated

**Comment by whitequark** [comment]
@dlrobertson Would be great!

**Comment by dlrobertson** [comment]
:+1: Added #152 which improves icmpv6 support and updated the description here.

**Comment by whitequark** [comment]
@m-labs-homu r+

**Comment by m-labs-homu** [comment]
:pushpin: Commit f6c5109 has been approved by `whitequark`

**Comment by m-labs-homu** [comment]
:hourglass: Testing commit f6c510920132cda1e3f952707f4c9ed0d9ae4058 with merge f06af168b4987ef8820d68f8a82d8b4c68e04cfc...

**Comment by whitequark** [comment]
Hm, autosquash seems broken.

**Comment by whitequark** [comment]
@m-labs-homu r-

**Comment by m-labs-homu** [comment]
:sunny: Test successful - [status-travis](https://travis-ci.org/m-labs/smoltcp/builds/337619575?utm_source=github_status&utm_medium=notification)
State: approved= try=False

**Pull request merged**